src/utils/allocators.zig

```diff
@@ -178,120 +192,154 @@ pub const DiskMemoryAllocator = struct {
-                return null;
-            };
-        }
+        };
 
-        const memory = std.posix.mmap(
+        const full_alloc = std.posix.mmap(
             null,
             aligned_size,
             std.posix.PROT.READ | std.posix.PROT.WRITE,
             std.posix.MAP{ .TYPE = .SHARED },
             file.handle,
             0,
         ) catch |err| {
-            std.debug.print("Disk Memory Allocator error: {}\n", .{err});
+            self.logFailure(err, file_name);
             return null;
         };
 
-        return memory.ptr;
-    }
-
-    /// unmaps the memory (file still exists and is removed on deinit())
-    pub fn free(_: *anyopaque, buf: []u8, log2_align: u8, return_address: usize) void {
-        _ = log2_align;
-        _ = return_address;
-        // TODO: build a mapping from ptr to file so we can delete the corresponding file on free
-        const buf_aligned_len = std.mem.alignForward(usize, buf.len, std.mem.page_size);
-        std.posix.munmap(@alignCast(buf.ptr[0..buf_aligned_len]));
+        std.mem.bytesAsValue(Metadata, full_alloc[size..][0..@sizeOf(Metadata)]).* = .{
+            .file_index = file_index,
+        };
+        return full_alloc.ptr;
     }
 
     /// not supported rn
     fn resize(
-        _: *anyopaque,
-        buf_unaligned: []u8,
-        log2_buf_align: u8,
+        ctx: *anyopaque,
+        buf: []u8,
+        log2_align: u8,
         new_size: usize,
         return_address: usize,
     ) bool {
-        // not supported
-        _ = buf_unaligned;
-        _ = log2_buf_align;
-        _ = new_size;
         _ = return_address;
-        return false;
+        const self: *Self = @ptrCast(@alignCast(ctx));
+
+        const alignment = @as(usize, 1) << @intCast(log2_align);
+        std.debug.assert(alignment <= std.mem.page_size); // the allocator interface shouldn't allow this (aside from the *Raw methods).
+
+        const buf_ptr: [*]align(std.mem.page_size) u8 = @alignCast(buf.ptr);
+        const aligned_size = fullMmapSize(buf.len);
+        const new_aligned_size = fullMmapSize(new_size);
+
+        const metadata: Metadata = @bitCast(buf_ptr[buf.len..][0..@sizeOf(Metadata)].*);
+        const file_name_bounded = fileNameBounded(metadata.file_index);
+        const file_name = file_name_bounded.constSlice();
+
+        if (aligned_size == new_aligned_size) return true;
+
+        if (new_aligned_size < aligned_size) {
+            std.posix.munmap(@alignCast(buf_ptr[new_aligned_size..aligned_size]));
+            std.mem.bytesAsValue(Metadata, buf_ptr[new_size..][0..@sizeOf(Metadata)]).* = .{
+                .file_index = metadata.file_index,
+            };
+            return true;
+        } else {
+            const file = self.dir.openFile(file_name, .{ .mode = .read_write }) catch |err| {
+                self.logFailure(err, file_name);
+                return false;
+            };
+            defer file.close();
+
+            const mapped = std.posix.mmap(
```

Comment:
Could this call to mmap be avoided if the initial mmap is sufficiently large? The initial alloc could create a small file while mmapping a much larger address space. For example, what about this:

1. alloc is called requesting less than one page of memory
2. alloc creates a file sized as 1 page
3. alloc mmaps 10 pages of address space, backed by the newly created file
4. alloc returns a slice with a len for the requested size (within the first page, which is valid memory)
5. resize is called requesting a size between 2 and 3 pages
6. the file is resized to be 3 pages in size
7. the same pointer can be reused without another call to mmap

The downside is the complexity of separately keeping track of the mmap size that was used during the prior call to alloc or resize, perhaps by storing it in the metadata.

Reply:
This is an interesting plan, and I think it could actually done without any extra state tracking, ie by using a fixed algorithm which calculates a requested length into a number of pages/page-size-compatible length.
Though this altogether also comes with the drawback of using more (disk) memory. A different tradeoff we could make would be to instead use DMA as the backing allocator of something like an arena allocator, or some other intelligent allocator which allocates larger amounts of memory at a time, to reduce the extent to which DMA has to do complicated tracking of this kind of state. Lmk your thoughts.

Reply:
> comes with the drawback of using more (disk) memory

My suggestion was to use the same amount of disk memory, not to increase it. I think you're talking about something different. The file size and mmap length are two distinct dimensions. Here are three strategies for managing those numbers:
- _the current implementation:_ the file size and the mmap length are the same number. both are the minimum number of pages to satisfy the alloc
- _what I think you're talking about when you say "more disk memory":_ the file size and the mmap length are the same number. this is an arbitrarily large number containing extra pages that are not needed now but can be used later.
- _what I was suggesting:_ the file and the mmap have different sizes from each other. the file size is the minimum number of pages to satisfy the alloc. the mmap length is a larger number that could support more pages being added to the file.

So, in my suggestion, the file would have the same size as your current approach. The only difference I was suggesting was to pass an arbitrarily large `length` argument into mmap. That means mmap would assign a larger address space to the pointer, but it wouldn't actually consume any extra memory in disk or ram. The full address space would not be usable at first, but that's ok because we only need a subset of that address space to satisfy the current alloc request.

The goal was just to avoid extra calls to mmap. You would still need to increase the size of the file for every resize. As long as the actual size of the file doesn't exceed the size of the address space created from the prior call to mmap, you could continue using the same pointer without needing to call mmap again.

> A different tradeoff we could make would be to instead use DMA as the backing allocator of something like an arena allocator, or some other intelligent allocator which allocates larger amounts of memory at a time, to reduce the extent to which DMA has to do complicated tracking of this kind of state. Lmk your thoughts.

This sounds like a good idea.

Reply:
Ah, I had a partial misunderstanding. I understood you intended for the file size and address space sizes to be different, but I actually didn't realize that mmapping more memory than exists doesn't consume disk space - in retrospect that's obvious, my bad.

As for the DMA as a backing allocator, mayhaps we can make the DMA configurable: a mode where it acts like it currently does, and a mode where it uses the slightly more intelligent strategy you've described that I now better comprehend the implications of.
I'll go ahead and do it this way, and you can tell me how you feel about each version, and if we keep both or just one.

Reply:
If you want to make it configurable, the configuration could be an integer parameter that specifies the ratio of mmap length to file size. Then it would essentially always be in the same mode of `mmap length >= file size`.  `mmap length == file size` would be a special case of the data, but would use the same logic.

Reply:
I actually attempted to do this, but I ran into an error where it appears that resizing the file invalidates the mapped pointer? It leads segmentation faults and bus errors when trying to store the metadata.
I'm going to eschew the more advanced version of this.

Reply:
ae191ee9c6db6640e068f82dda492c766ddf8b80